src/compiler/utilities.ts

```diff
@@ -5963,6 +5963,22 @@ namespace ts {
         return option.strictFlag ? getStrictOptionValue(options, option.name as StrictOptionName) : options[option.name];
     }
 
+    export function getJSXTransformEnabled(options: CompilerOptions): boolean {
+        const jsx = options.jsx;
+        return jsx === JsxEmit.React || jsx === JsxEmit.ReactJSX || jsx === JsxEmit.ReactJSXDev;
+    }
+
+    export function getJSXImplicitImportBase(compilerOptions: CompilerOptions, file: SourceFile): string | undefined {

```

Comment:
Hi! @weswigham I think this is wrong. You should not treat it as a "base" of the import. It should be a fully qualified import specific. With the current behavior, it is impossible to make it emit like: `import { jsx as _jsx } from "//cdn.example.com/my-jsx-lib/jsx-runtime.js"` cause it is not possible to add a ".js" at the end of the import specifier.

This also requires the imported file to name exactly as "jsx-runtime".

Reply:
...Yeah, so that wasn't _really_ my choice - it's how the `react` maintainers implemented the same option in [the `babel` transform.](https://babeljs.io/docs/en/babel-preset-react#importsource) (And we just want to match the functionality provided by babel here, without overstepping.) The runtime being accessible at `${source}/jsx-runtime` and `${source}/jsx-dev-runtime`, based on how the options are presented, is part of the API contract for these endpoints (probably so the import can be swapped between dev/non-dev without explicitly reconfiguring the import? I dunno, I'm guessing.).

Reply:
Thanks for your explanation, I'll post it to the react repo instead

Reply:
facebook/react#19905

Reply:
hi @weswigham babel/babel#12116 babel has added `.js` after `jsx-runtime`. Please also do that in TypeScript, thanks!